File: core/src/main/scala/com/graphbrain/eco/nodes/VertexFun.scala
package com.graphbrain.eco.nodes

import com.graphbrain.eco.NodeType.NodeType
import com.graphbrain.eco.{Contexts, NodeType}
import com.graphbrain.db.{ID, EntityNode}

class VertexFun(val fun: VertexFun.VertexFun, params: Array[ProgNode], lastTokenPos: Int= -1) extends FunNode(params, lastTokenPos) {

  override val label = fun match {
    case VertexFun.BuildVert => "build"
    case VertexFun.RelVert => "rel-vert"
    case VertexFun.TxtVert => "txt-vert"
  }

  override def ntype: NodeType = NodeType.Vertex

  override def vertexValue(ctxts: Contexts) = {
    fun match {
      case VertexFun.BuildVert => {
        for (p <- params) {
          p.ntype match {
            case NodeType.Vertex => p.vertexValue(ctxts)
            case NodeType.String => p.stringValue(ctxts)
            case _ => "" // error!
          }
        }

        for (c <- ctxts.ctxts) {
          val id = "(" + params.map(
            p => p.ntype match {
              case NodeType.Vertex => c.getRetVertex(p)
              case NodeType.String => c.getRetString(p)
              case _ => "" // error!
            }
          ).reduceLeft(_ + " " + _) + ")"

          c.setRetVertex(this, id)
        }
      }
      case VertexFun.RelVert => {
        val p = params(0)

        p.ntype match {
          case NodeType.Words => p.wordsValue(ctxts)
          case NodeType.String => p.stringValue(ctxts)
          case _ => "" // error!
        }

        for (c <- ctxts.ctxts) {
          p.ntype match {
            case NodeType.Words => c.setRetVertex(this, ID.reltype_id(c.getRetWords(p).text))
            case NodeType.String => c.setRetVertex(this, ID.reltype_id(c.getRetString(p)))
            case _ => "" // error!
          }
        }
      }
      case VertexFun.TxtVert => {
        val p = params(0)

        p.ntype match {
          case NodeType.Words => p.wordsValue(ctxts)
          case NodeType.String => p.stringValue(ctxts)
          case _ => "" // error!
        }

        for (c <- ctxts.ctxts) {
          p.ntype match {
            case NodeType.Words => c.setRetVertex(this, EntityNode.id("1", c.getRetWords(p).text))
            case NodeType.String => c.setRetVertex(this, EntityNode.id("1", c.getRetString(p)))
            case _ => "" // error!
          }
        }
      }
    }
  }

  override protected def typeError() = error("parameters must be variables or strings")
}

object VertexFun extends Enumeration {
  type VertexFun = Value
  val BuildVert,
    RelVert,
    TxtVert = Value
}

File: core/src/main/java/com/graphbrain/eco/nodes/VertexFun.java
package com.graphbrain.eco.nodes;

import com.graphbrain.db.*;
import com.graphbrain.eco.Context;
import com.graphbrain.eco.Contexts;
import com.graphbrain.eco.NodeType;

public class VertexFun extends FunNode {

    public enum VertexFunType {
        BuildVert, RelVert, TxtVert, Flatten
    }

    private VertexFunType fun;

    public VertexFun(VertexFunType fun, ProgNode[] params, int lastTokenPos) {
        super(params, lastTokenPos);
        this.fun = fun;
    }

    @Override
    public String label() {
        switch(fun) {
            case BuildVert: return "build";
            case RelVert: return "rel-vert";
            case TxtVert: return "txt-vert";
            case Flatten: return "flatten";
            default: return "?";
        }
    }

    @Override
    public NodeType ntype(Context ctxt) {return NodeType.Vertex;}

    @Override
    public void eval(Contexts ctxts) {
        switch(fun) {
            case BuildVert:
                for (ProgNode p : params) {
                    p.eval(ctxts);
                }

                for (Context c : ctxts.getCtxts()) {
                    Vertex[] verts = new Vertex[params.length];
                    for (int i = 0; i < params.length; i++) {
                        ProgNode p = params[i];
                        switch(p.ntype(c)) {
                            case Vertex:
                                verts[i] = c.getRetVertex(p);
                                break;
                            case String:
                                verts[i] = Vertex.fromId(c.getRetString(p));
                                break;
                            default:
                                // error
                                break;
                        }
                    }
                    c.setRetVertex(this, new Edge(verts));
                }
                break;
            case RelVert:
                ProgNode p = params[0];

                p.eval(ctxts);

                for (Context c : ctxts.getCtxts()) {
                    switch(p.ntype(c)) {
                        case Words:
                            c.setRetVertex(this, Vertex.fromId(EdgeType.buildId(c.getRetWords(p).text())));
                            break;
                        case String:
                            c.setRetVertex(this, Vertex.fromId(EdgeType.buildId(c.getRetString(p))));
                            break;
                        default: // error!
                    }
                }
                break;
            case TxtVert: {
                p = params[0];

                p.eval(ctxts);

                for (Context c : ctxts.getCtxts()) {
                    switch(p.ntype(c)) {
                        case Words:
                            c.setRetVertex(this, EntityNode.fromNsAndText("", c.getRetWords(p).text()));
                            break;
                        case String:
                            c.setRetVertex(this, EntityNode.fromNsAndText("", c.getRetString(p)));
                            break;
                        default: // error!
                    }
                }
                break;
            }
            case Flatten: {
                p = params[0];
                p.eval(ctxts);

                for (Context c : ctxts.getCtxts()) {
                    Vertex v = c.getRetVertex(p);
                    c.setRetVertex(this, v.flatten());
                }
                break;
            }
        }
    }
}